### PR TITLE
fix: on_response callbacks now spawned as tasks

### PR DESCRIPTION
**Pull request by hpopp**
Expensive callbacks can actually block the worker (as they are processed sequentially). New implementation also has the benefit of not crashing the worker if the callback causes trouble. Unsure why this wasn't implemented properly in the first place.

@talklittle This can be deployed as a bugfix release, but I'd like a second opinion as to whether there might be unintended side effects. Also need a sanity check on the ADM worker changes. I'm a little confused about this block. It looks like it's still using the old `{:error, reason, notif}` response style.

```elixir
  # Pigeon.ADM.Worker
  def handle_cast({:push, :adm, notification, on_response}, state) do
    case refresh_access_token_if_needed(state) do
      {:ok, state} ->
        :ok = do_push(notification, state, on_response)
        {:noreply, state}
      {:error, reason} ->
        on_response.({:error, reason, notification})
        {:noreply, state}
    end
  end
```

### CODE REVIEW

File: lib/pigeon/adm.ex [hunk 1]
```diff
@@ -120,9 +120,9 @@ defmodule Pigeon.ADM do
   defp cast_push(worker_name, notifications, on_response) when is_list(notifications) do
     for n <- notifications, do: cast_push(worker_name, n, on_response)
   end
-  defp cast_push(worker_name, notification, nil) do
-    GenServer.cast(worker_name, {:push, :adm, notification})
-  end
+  #defp cast_push(worker_name, notification, nil) do
+  #  GenServer.cast(worker_name, {:push, :adm, notification})
+  #end
```
Comment (hpopp) [inline lib/pigeon/adm.ex]: I don't *think* this is used anymore, but not sure.

Comment (talklittle) [reply lib/pigeon/adm.ex]: Looks like you are correct, can be deleted.

Comment (talklittle) [reply lib/pigeon/adm.ex]: Remember to also delete the corresponding unused clause of `handle_cast/2` in ADM.Worker line 38

**Comment by talklittle** [comment]
IMO this PR is a breaking change since it changes response callbacks to parallel instead of single-threaded, and that might be unexpected. Before, callbacks would fire in the same order as they were supplied to Pigeon; after this change, that's no longer guaranteed. So the Pigeon version number should be increased accordingly (1.2.0 instead of 1.1.2).

As for that line you pointed out in ADM.Worker, that is probably a bug, and I can submit a separate PR to fix it. That one is probably safe as a bugfix, since there's not much an application can do in case the token refresh fails, and even crashing the worker here would not be catastrophic.

**Comment by hpopp** [comment]
Order isn't guaranteed in async callbacks already. It isn't called until kadabra receives an END_STREAM for that request, which can happen out of order (and frequently does).

Ultimately its a matter of moving the callback work from the Worker to a new process, which shouldn't actually change the end result. I agree with you that this seems like a breaking change, but I still can't entirely justify why. This is a somewhat serious bug that should be fixed soon, though I would also like to have more changes for a `v1.2` release.

Go ahead and set up a PR for cleaning up ADM as you see fit.

**Comment by talklittle** [comment]
> Order isn't guaranteed in async callbacks already. It isn't called until kadabra receives an END_STREAM for that request, which can happen out of order (and frequently does).

Good point. Thanks for the clarification. Maybe then it's only ADM that will have the parallel callbacks behavior change, since it's still using HTTPoison.

Now that I understand the FCM/APNS callbacks had been running in the Connection worker process, I can't think of a downside to "letting go" of the HTTP2 connection process sooner, and sending the `on_response` to another process to be executed. Since the `on_response` should not care at all about the current network state. So :+1:

**Comment by hpopp** [comment]
After reading through hackney's source I'm pretty sure its async responses don't guarantee ordering either. I'm going to go ahead and include ADM in this fix and release as `1.1.4`

**Comment by talklittle** [comment]
My bad, I was looking at the `refresh_access_token/1` which uses HTTPoison synchronously, rather than `do_push/3` which already creates a new Task per HTTPoison request. You're right again, of course.

**Pull request merged**